Split read recovery memory equally between partitions

mmaslankaprv opened the pull request:
Previously Raft recovery was configured to use static max read buffer
size of `512 KiB`. This may make recovery slow especially if there is
only few partitions recovering per shard as they will not use the shard
recovery memory effectively.

Changed the implementation to use a 1/max_allowed_concurrent_recoveries.
This way a read buffer will be larger and we will still not starve
any partitions recovering as there is maximum number of recoveries
running on the shard.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
### Improvements
- Faster partition movements in low partition settings

mmaslankaprv committed:
r/recovery_mem: remove configuration provider function

Removed configuration provider function and using
`seastar::sharded_parameter` instead to pass in the shard local
configuration bindings.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
r/recovery_mem: use all available memory when reading for recovery

Previously Raft recovery was configured to use static max read buffer
size of 512 KiB. This may make recovery slow especially if there is
only few partitions recovering per shard as they will not use the shard
recovery memory effectively.

Changed the implementation to use a 1/max_allowed_concurrent_recoveries.
This way a read buffer will be larger and we will still not starve
any partitions recovering as there is maximum number of recoveries
running on the shard.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -432,10 +432,12 @@ configuration::configuration()
       "raft_recovery_default_read_size",
       "Specifies the default size of a read issued during Raft follower "
-      "recovery.",
+      "recovery. (by default this property is disabled and max recovery read "
```
I also wonder whether the sentence. "By default... should go in a note as for [https://docs.redpanda.com/current/reference/properties/cluster-properties/#retention_local_target_bytes_default](https://docs.redpanda.com/current/reference/properties/cluster-properties/#retention_local_target_bytes_default). 

```suggestion
      "recovery. By default, this property is disabled and the maximum recovery read "
```

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -432,10 +432,12 @@ configuration::configuration()
       "Specifies the default size of a read issued during Raft follower "
-      "recovery.",
+      "recovery. (by default this property is disabled and max recovery read "
+      "size is calculated as a result of "
```
```suggestion
      "size is calculated as "
```

asimms41 reviewed (commented):
Just some suggestions.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--83b04d46-2485-4acc-a12f-1945f0ec8f76-->
<!--TEST RESULTS-->
<!--0a16127d-7dbc-4d93-8a95-b339e99cdbfb-->
<!--TEST RESULTS-->
<!--a5d0a45c-5276-43ce-8d48-318635d28713-->
<!--TEST RESULTS-->
<!--efadb4ab-cdb1-4dd0-b3b9-bb51c8a781f9-->
<!--TEST RESULTS-->
<!--9f704da0-6374-41bd-87a4-47679d473880-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71684">build#71684</a></summary>


| test_class            | test_method                        | test_arguments                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:----------------------|:-----------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_migrated_topic_data_integrity | {"include_groups": false, "params": {"cancellation": {"dir": "in", "stage": "executed"}, "use_alias": false}, "transfer_leadership": true} | integration | https://buildkite.com/redpanda/redpanda/builds/71684#019904c0-b44f-4d4f-bcf8-85e2c2928da9 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS             |
| DatalakeE2ETests      | test_json_schema_unicode           | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/71684#019904c0-b44f-4d4f-bcf8-85e2c2928da9 | FLAKY         | 20/21    | upstream reliability is '99.46949602122017'. current run reliability is '95.23809523809523'. drift is 4.2314 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71978">build#71978</a></summary>


| test_class                | test_method              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:--------------------------|:-------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| test_task_started_fixture | test_change_run_interval |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/71978#0199333c-1f01-44f6-9c66-69c4efb0c16f | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72083">build#72083</a></summary>


| test_class        | test_method            | test_arguments                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------|:-----------------------|:-----------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTest | compaction_stress_test | {"cleanup_policy": "compact,delete", "key_set_cardinality": 100, "storage_compaction_key_map_memory_kb": 131072} | integration | https://buildkite.com/redpanda/redpanda/builds/72083#01993936-fa25-4d19-9682-2477c70a39c5 | FLAKY         | 20/21    | upstream reliability is '99.43181818181817'. current run reliability is '95.23809523809523'. drift is 4.19372 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72132">build#72132</a></summary>


| test_class               | test_method                          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:-------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_creating_and_listing_migrations | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72132#01993cae-994d-4aa8-a04a-43fd3845eaa0 | FLAKY         | 20/21    | upstream reliability is '99.16666666666667'. current run reliability is '95.23809523809523'. drift is 3.92857 and the allowed drift is set to 50. The test should PASS  |
| PartitionBalancerTest    | test_rack_awareness                  | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72132#01993cae-994f-4289-992b-f1bc825df204 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest | test_node_operations                 | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72132#01993cae-9951-495f-b126-f856be3fe39d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| RandomNodeOperationsTest | test_node_operations                 | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/72132#01993cae-994b-4148-8a2e-6f701d373662 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              |
| SimpleEndToEndTest       | test_consumer_interruption           | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72132#01993cab-f768-4c95-aec3-09e727e5c5d1 | FLAKY         | 18/21    | upstream reliability is '99.29824561403508'. current run reliability is '85.71428571428571'. drift is 13.58396 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72328">build#72328</a></summary>


| test_class            | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest | test_recovery_mode_rebalance_finish | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72328#01995202-c832-4717-bb1d-07eb05606fe3 | FLAKY         | 15/21    | upstream reliability is '83.86167146974063'. current run reliability is '71.42857142857143'. drift is 12.4331 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

mmaslankaprv commented:
/cdt

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -432,10 +432,12 @@ configuration::configuration()
       "raft_recovery_default_read_size",
       "Specifies the default size of a read issued during Raft follower "
-      "recovery.",
+      "recovery. By default this property is disabled and the maximum recovery "
```
```suggestion
      "recovery. By default, this property is disabled and the maximum recovery "
```

asimms41 reviewed (commented):
Minor suggestion.

joe-redpanda commented on src/v/raft/recovery_memory_quota.cc:
```diff
@@ -20,23 +20,30 @@
 void recovery_memory_quota::on_max_memory_changed() {
-    int64_t new_size = int64_t(_cfg.max_recovery_memory().value_or(
-      memory_groups().recovery_max_memory()));
+    int64_t new_size = int64_t(
```
Do we want a boundary check on size_t -> int64_t

joe-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -432,10 +432,12 @@ configuration::configuration()
       {.needs_restart = needs_restart::no, .visibility = visibility::tunable},
-      512_KiB,
-      {.min = 128, .max = 5_MiB})
+      std::numeric_limits<int32_t>::max(),
```
Can we keep some reasonable max on this?

bharathv replied:
why not deprecate this? I thought that was the original idea.

mmaslankaprv replied:
yes, but there is a lot of code pieces that rely on this to be presents, tests especially

bharathv replied:
didn't fully follow.. I only see it is used in compaction_e2e_test.py?

mmaslankaprv replied:
actually you are right, let me deprecate that one

joe-redpanda commented on src/v/raft/recovery_memory_quota.cc:
```diff
@@ -20,23 +20,30 @@
-      std::min(_current_max_recovery_mem, _cfg.default_read_buffer_size()));
+    auto size = std::min(
+      _default_read_buffer_size(),
+      _max_recovery_memory().value_or(memory_groups().recovery_max_memory())
```
q: where is memory_groups getting its max recovery memory from?

mmaslankaprv replied:
it is calculated based on the total available memory on shard

mmaslankaprv committed:
config:deprecate `raft_recovery_default_read_size` property

The property is now deprecated as we are deriving the read size from the
recovery concurrency limit.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/ci-repeat 1

bharathv commented on src/v/raft/recovery_memory_quota.cc:
```diff
@@ -20,23 +20,28 @@
-    return ss::get_units(
-      _memory,
-      std::min(_current_max_recovery_mem, _cfg.default_read_buffer_size()));
+    auto size = static_cast<double>(_max_recovery_memory().value_or(
+                  memory_groups().recovery_max_memory()))
+                / static_cast<double>(_raft_recovery_concurrency_per_shard());
```
wondering if a minimum/max read size should be enforced.. something like..

min(max(512KiB, std::ceil(size), 5MiB);

mmaslankaprv replied:
maybe something larger like `32MiB` ?

bharathv replied:
oops yes, didn't mean 5MiB to be exact.. 32MiB makes sense to me.

bharathv replied:
@mmaslankaprv should we also enforce a minimum value (512KB in the example above)? If the tunables are configured such that each recovery gets only a tiny amount of memory, then all recoveries could suffer due to insufficient allocation.

bharathv reviewed (commented):
lgtm one question

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-27428-v25.2.x-720 remotes/upstream/v25.2.x
git cherry-pick -x 780710f4ef aceb27d66d 6a1d9a98e6
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/17773405850).
